src/nuclide.cpp

```diff
@@ -90,9 +90,21 @@ Nuclide::Nuclide(hid_t group, const std::vector<double>& temperature)
   double T_min = n > 0 ? settings::temperature_range[0] : 0.0;
   double T_max = n > 0 ? settings::temperature_range[1] : INFTY;
   if (T_max > 0.0) {
-    for (auto T : temps_available) {
-      if (T_min <= T && T <= T_max) {
-        temps_to_read.push_back(std::round(T));
+    // For each interval (T_j, T_j+1), if either T_j or T_j+1 are within the
```

Comment:
Shouldn't we be able to remove the conditional `if (T_max > 0.0)`? I would rather have `assert(T_max > 0.0)` above this. Or `gsl`'s `Expects`, which is the same thing as far as I can tell.

Personally I'd rather us move away from GSL since we only end up using `index` and `span` which are like 10 lines of code to implement ourselves, assuming we're not hoping to handle C-style fixed length arrays with a span implementation (which we don't use with spans regardless).

This has mainly been on my mind in trying to remove some warnings when compiling with `-Wall` and worrying about compatibility with accelerators (`assert` is in CUDA, but I've not tested GSL for device code). I think the benefit of enabling developers to turn on compiler warnings outweighs the benefits of GSL. The two are not necessarily mutually exclusive, but given that `gsl::index` is signed, you can't get away from all the sign comparison warnings that pop up. IMO it's nicer to put `vector<whatever>::size_type` or typedef that in a class for shorthand so you know for a fact the vector's size_type matches that of the index. The takeaway point here being that GSL will give us warnings due to using a signed integer index, so we should get rid of GSL and use `assert` rather than `Expects`. As further reason to do so, just do ctrl-f on "???" on the GSL webpage. Yikes.

If you can't tell, I'm trying to get rid of all our compiler warnings. There are many, most of which being due to using `int` to compare against `size_t`.

Sorry for the wall of text. Feel I had to motivate the opinion on why `assert` should replace `Expects`.

Reply:
A couple of thoughts:  
* gsl-lite lets you configure `gsl::index` to be either a `ptrdiff_t` (signed) or a `size_t` (unsigned):

https://github.com/gsl-lite/gsl-lite/blob/503b14bdd1cfa9a797dc3dddab6baac7e13504bf/include/gsl/gsl-lite.hpp#L223-L236

* I think signed loop counters have a better-informed consensus these days.  NVIDIA even recommends signed loop counters for CUDA code (see [this part of the CUDA best practices guide](https://docs.nvidia.com/cuda/cuda-c-best-practices-guide/index.html#loop-counters-signed-vs-unsigned)).

Reply:
Ah, really interesting on the second note @RonRahaman! I'll have to modify my code some to use that. Either way, maybe we should just define our own `index` type. No need to add a dependency just to get a typedef.

Reply:
The discussion on gsl and indexing is not really relevant to this PR, so I'm going to ignore it. Please feel free to open an issue for discussion, or post a message on the discourse forum (our development topic there is lonely anyway).

Regarding `T_max > 0.0`, the default value for `temperature_range` is {0.0, 0.0}, so this check really indicates that the user has manually specified a temperature range. When no range is given at all, `T_max` will be zero.

Reply:
> Sorry, just one more comment on Gavin's thoughts:
> 
> > I would rather have assert(T_max > 0.0) above this. Or gsl's Expects, which is the same thing as far as I can tell.
> 
> `assert` is actually ifdef'ed out when the `NDEBUG` preprocessor macro is defined (see [here](https://www.cplusplus.com/reference/cassert/assert/)). Likewise, CMake's ~"Debug"~ "Release" configuration adds `-DNDEBUG` to the compiler line (not sure if OpenMC disables that). So `assert` isn't appropriate for something that is essential to the correct functioning of the code.

Ah, yes, it seemed to me this was just a check that the minimum temperature is positive, not for correctness. I think the setup as-is is a bit hard to glean that this is for correctness rather than error checking.